src/Reflection/InitializerExprTypeResolver.php

```diff
@@ -1453,23 +1453,23 @@ private function callOperatorTypeSpecifyingExtensions(Expr\BinaryOp $expr, Type
 	 */
 	private function resolveCommonMath(Expr\BinaryOp $expr, Type $leftType, Type $rightType): Type
 	{
-		if (($leftType instanceof IntegerRangeType || $leftType instanceof ConstantIntegerType || $leftType instanceof UnionType) &&
-			($rightType instanceof IntegerRangeType || $rightType instanceof ConstantIntegerType || $rightType instanceof UnionType)
-		) {
+		$leftNumberType = $leftType->toNumber();
+		$rightNumberType = $rightType->toNumber();
 
-			if ($leftType instanceof ConstantIntegerType) {
+		if ($rightNumberType instanceof IntegerRangeType || $rightNumberType instanceof ConstantIntegerType || $rightType instanceof UnionType) {
+			if ($leftNumberType instanceof IntegerRangeType || $leftNumberType instanceof ConstantIntegerType) {
 				return $this->integerRangeMath(
-					$leftType,
+					$leftNumberType,
 					$expr,
-					$rightType,
+					$rightNumberType,
 				);
 			} elseif ($leftType instanceof UnionType) {
-
 				$unionParts = [];
 
 				foreach ($leftType->getTypes() as $type) {
-					if ($type instanceof IntegerRangeType || $type instanceof ConstantIntegerType) {
-						$unionParts[] = $this->integerRangeMath($type, $expr, $rightType);
+					$numberType = $type->toNumber();
+					if ($numberType instanceof IntegerRangeType || $numberType instanceof ConstantIntegerType) {
```

Comment:
Would this work with `isInteger()->yes()` instead of `instanceof`?

Reply:
We don't want `instanceof Integer' but only IntegerRange and ConstantInteger.